pkg/test/context.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"log"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"k8s.io/client-go/rest"
+)
+
+type TestCtx struct {
+	ID         string
+	CleanUpFns []finalizerFn
+	Namespace  string
+	CRClient   *rest.RESTClient
+}
+
+type finalizerFn func() error
+
+func (f *Framework) NewTestCtx(t *testing.T) TestCtx {
+	var prefix string
+	if t != nil {
+		// TestCtx is used among others for namespace names where '/' is forbidden
+		prefix = strings.TrimPrefix(
+			strings.Replace(
+				strings.ToLower(t.Name()),
+				"/",
+				"-",
+				-1,
+			),
+			"test",
+		)
+	} else {
+		prefix = "main"
+	}
+
+	id := prefix + "-" + strconv.FormatInt(time.Now().Unix(), 10)
+	return TestCtx{
+		ID: id,
+	}
+}
+
+func (ctx *TestCtx) GetID() string {
+	return ctx.ID
+}
+
+// GetObjID returns an ascending ID based on the length of cleanUpFns. It is
+// based on the premise that every new object also appends a new finalizerFn on
+// cleanUpFns.
+func (ctx *TestCtx) GetObjID() string {
+	return ctx.ID + "-" + strconv.Itoa(len(ctx.CleanUpFns))
+}
+
+func (ctx *TestCtx) Cleanup(t *testing.T) {
```

Comment:
Can you have make `Cleanup` to take in `f func(format string, v ...interface{})`?
then you can pass `t.Errorf` or `log.Fatalf` to it. So that we don't have to have a `CleanupNoT` function.

e.g:

```go
func (ctx *TestCtx) Cleanup(f func(format string, v ...interface{})) {
	for i := len(ctx.CleanUpFns) - 1; i >= 0; i-- {
		err := ctx.CleanUpFns[i]()
		if err != nil {
			f("a cleanup function failed with error: %v\n", err)
		}
	}
}

func TestBla(t *testing.T) {
 ctx := Global.NewTestCtx(nil)
 defer ctx.Cleanup(log.Fatalf)

 ctx = Global.NewTestCtx(t)
 defer ctx.Cleanup(t.Errof)
}
```

Reply:
@AlexNPavel any objection to the above approach?

Reply:
I like it. What I'm not sure about is how to handle failure. We don't want to use `log.Fatalf` as our error reporting function, because that could cause the function to end before all clean up functions have finished. The way that `CleanupNoT` handles this is with a `failed` boolean that only runs `log.Fatalf` after all functions have been executed. We could use that design again, but there's still the issue about what happens if a user give us something like `log.Fatalf`. We could document the behavior, but it would still be easy for people to make this mistake.

Reply:
I see. Let's just keep it is at the moment. we can revisit this later.